types/src/validator_change.rs

```diff
@@ -69,32 +88,69 @@ impl ValidatorChangeProof {
-    /// In case waypoint is present it's going to be used for verifying the very first epoch change
-    /// (it's the responsibility of the caller to not pass waypoint in case it's not needed).
+    /// Verify the proof is correctly chained with known epoch and validator
+    /// verifier and return the [`LedgerInfoWithSignatures`] to start target epoch.
+    ///
+    /// In case a waypoint is present, it's going to be used for verifying the
+    /// very first epoch change (it's the responsibility of the caller to not
+    /// pass a waypoint in case it's not needed).
+    ///
+    /// We will also skip any stale ledger info's in the [`ValidatorChangeProof`].
     pub fn verify(&self, verifier: &VerifierType) -> Result<LedgerInfoWithSignatures> {
         ensure!(
             !self.ledger_info_with_sigs.is_empty(),
-            "Empty ValidatorChangeProof"
+            "The ValidatorChangeProof is empty"
         );
+        ensure!(
+            !verifier
+                .is_ledger_info_stale(self.ledger_info_with_sigs.last().unwrap().ledger_info()),
+            "The ValidatorChangeProof is stale as our verifier is already ahead \
+                of the entire ValidatorChangeProof"
+        );
+
         self.ledger_info_with_sigs
             .iter()
-            .try_fold(verifier.clone(), |verifier, ledger_info| {
-                verifier.verify(ledger_info)?;
-                // While the original verification could've been via waypoints, all the next epoch
-                // changes are verified using the (already trusted) validator sets.
-                ledger_info
+            // Skip any stale ledger infos in the proof prefix. Note that with
+            // the assertion above, we are guaranteed there is at least one
+            // non-stale ledger info in the proof.
+            //
+            // It's useful to skip these stale ledger infos to better allow for
+            // concurrent client requests.
+            //
+            // For example, suppose the following:
+            //
+            // 1. My current trusted state is at epoch 5.
+            // 2. I make two concurrent requests to two validators A and B, who
+            //    live at epochs 9 and 11 respectively.
+            //
+            // If A's response returns first, I will ratchet my trusted state
+            // to epoch 9. When B's response returns, I will still be able to
+            // ratchet forward to 11 even though B's ValidatorChangeProof
+            // includes a bunch of stale ledger infos (for epochs 5, 6, 7, 8).
+            //
+            // Of course, if B's response returns first, we will reject A's
+            // response as it's completely stale.
+            .skip_while(|&ledger_info_with_sigs| {
+                verifier.is_ledger_info_stale(ledger_info_with_sigs.ledger_info())
+            })
+            // Try to verify each (epoch -> epoch + 1) jump in the ValidatorChangeProof.
```

Comment:
Do you think we should fail verification in case the whole list is stale? I think it doesn't make sense to allow stale chains (one way to do that would be to check that the very last member is stale prior the iter loop...).

Reply:
Ahh it does fail if the whole chain is stale. From this check:

```rust
        ensure!(
            !verifier
                .is_ledger_info_stale(self.ledger_info_with_sigs.last().unwrap().ledger_info()),
            "The ValidatorChangeProof is stale as our verifier is already ahead \
            of the entire ValidatorChangeProof"
        );
```

Reply:
Oops, sorry, missed that!